core/src/banking_stage.rs

```diff
@@ -547,151 +540,72 @@ impl BankingStage {
         recorder: &TransactionRecorder,
         qos_service: &QosService,
         slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+        num_packets_to_process_per_iteration: usize,
     ) {
         let mut rebuffered_packet_count = 0;
         let mut consumed_buffered_packets_count = 0;
-        let buffered_packet_batches_len = buffered_packet_batches.len();
+        let buffered_packets_len = buffered_packet_batches.len();
         let mut proc_start = Measure::start("consume_buffered_process");
         let mut reached_end_of_slot: Option<EndOfSlot> = None;
 
-        RetainMut::retain_mut(buffered_packet_batches, |deserialized_packet_batch| {
-            let original_unprocessed_indexes = deserialized_packet_batch
-                .unprocessed_packets
-                .keys()
-                .cloned()
-                .collect::<Vec<usize>>();
-            if let Some(end_of_slot) = &reached_end_of_slot {
-                // We've hit the end of this slot, no need to perform more processing,
-                // just filter the remaining packets for the invalid (e.g. too old) ones
-                // if the working_bank is available
-                let mut end_of_slot_filtering_time = Measure::start("end_of_slot_filtering");
-                let should_retain = if let Some(bank) = &end_of_slot.working_bank {
-                    let new_unprocessed_indexes = Self::filter_unprocessed_packets_at_end_of_slot(
-                        bank,
-                        &deserialized_packet_batch.unprocessed_packets,
-                        my_pubkey,
-                        end_of_slot.next_slot_leader,
-                        banking_stage_stats,
-                    );
+        // All retryable packets get inserted here
+        let mut retryable_packets =
+            UnprocessedPacketBatches::with_capacity(buffered_packet_batches.capacity());
```

Comment:
yeah agreed

> or flag the packet in buffer when sent down for processing, then remove it if execution was successful or un-flag if retryable

I thought about this and it was a possibility but decided it would make v1 more complicated

1. You could take an iterator through the heap, but the tradeoff here is you keep around already processed packets, so that saturates the buffer
2. Already processed packets will have to be iterated through every time
3. You eventually have to filter the min-max heap to get rid of the successfully processed packets.

I think the tradeoff is potentially worth it if a lot of the transactions are failing with retryable errors and you can skip the rebuffering, but that seems to be optimizing for a less common case.

Reply:
Mmm, I was thinking to removed processed packets at the end of each consuming loop, similar to current behavior.

I agree this can be in V2, given all the pef improvements done for single packets.